### DISCUSSION TITLE
Qual a estratégia profissional para realização de atividades frequentes automáticas

### DISCUSSION THREAD

**Discussion by EuHigorBarbosa** [Flutterando/forum#119] [Perguntas e Respostas]
Oi pessoal, tudo bem?

Uma dúvida...meu cliente deseja que o aplicativo faça um backup das informações locais no banco de dados a cada dois dias. 

Então é o caso de uma ação automática no aplicativo que vai enviar dados para o banco de dados. 

E aí que tá...eu não tenho experiência prévia para implementar esse requisito. 

Poderiam me sugerir uma estratégia profissional para implementar esse requisito?

Pesquisando eu vi alguns packages de schedule de tasks como o background_fetch (https://pub.dev/packages/background_fetch) que funciona tanto para IOS quanto para Android,  mas seria essa a estratégia adequada? 

Existe alguma mais adequada? 

Obrigado.

**Comment by eduardoflorence** [comment]
Oi @EuHigorBarbosa, seu aplicativo usa um banco de dados no dispositivo? Qual seria?

**Reply by EuHigorBarbosa** [reply]
Meu app usa o Firebase Firestore como banco de dados. Os dados deverão sair da memória local do celular para serem colocados no firebase (para servirem de backup) de dois em dois dias.

**Reply by EuHigorBarbosa** [reply]
Agora que percebi, vc perguntou se meu aplicativo usa um banco de dados no dispositivo...como um banco de dados local. A resposta é não. Eu gravo as informações no dispositivo por meio de funções que eu mesmo criei...com a biblioteca do Dart:io. Eu uso a função writeAsString...como mostrado abaixo.

```
@override
  Future<bool> saveStringJsonInLocal(
      {required String nomeArquivo,
      required String pathLocalMainFolder,
      required String dataStringJson}) async {
    if(await permissionManager.permissionStorageIsGranted()){
      try {
        /// transforma um List<Map<String, dynamic>> em string-json
        File fileToSave = File(pathLocalMainFolder + nomeArquivo);
        print(fileToSave.path);

        fileToSave.writeAsString(dataStringJson);
      } catch (e,s) {
        print(e);
        throw UnknowFailureLocalStorageManipulation(s,
            'Erro ao salvar dados no endereço $nomeArquivo usado pelo aplicativo', e, e.toString());
      }
      return true;
    } else {
      throw UnAuthorizedStoragePermissionError(null,
            'O aplicativo não tem permissão para salvar arquivos', 'Erro no salvamento de arquivos', 'Erro no salvamento de arquivos');
    }
  }```

Daí quando eu quero buscar o dado eu só indico o endereço string.

**Reply by eduardoflorence** [reply]
Por quê você não grava tudo diretamente no Firestore (Firebase)?
Se não houver Internet, as informações são gravadas automaticamente em cache e quando volta a Internet, elas são sincronizadas.
Vide: https://firebase.google.com/docs/firestore/manage-data/enable-offline?hl=pt-br

**Reply by EuHigorBarbosa** [reply]
Então, o aplicativo foi concebido com a seguinte ideia em mente...a de que o app faria o que ele foi feito pra fazer offline. E de tempos em tempos iria gerar um backup com algumas informações. Ele não foi construído de forma a enviar os dados para o banco de dados a cada modificação de dados dentro do App. Por isso precisava fazer um schedule de task de forma com que ele alimentasse o BD a cada 2 dias com algumas informações que permitiriam o usuário não perder os dados essenciais. 

O programa é um programa de desenvolvimento pessoal criado para terapeutas e alunos de terapeutas. Esse programa grava áudios offline e os concatena de forma a formar frases que fazem com que o usuário tenha benefícios psicológicos ao ouví-los. Essa geração é toda feita off line e nós não armazenamos todos os audios no BD. O intuito do backup não é gravar os áudios mas sim gravar as informações que possibilitem o usuário regravar os audios. 

Por isso não posso enviar os dados para o Firebase pois teria que fazer uma grande modificação no programa.
